src/tools/illink/src/linker/Linker.Steps/MarkStep.PropertyDefinitionNode.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using ILCompiler.DependencyAnalysisFramework;
+using Mono.Cecil;
+
+namespace Mono.Linker.Steps
+{
+
+	public partial class MarkStep
+	{
+		internal sealed class PropertyDefinitionNode : DependencyNodeCore<NodeFactory>
+		{
+			PropertyDefinition _property;
+			public PropertyDefinitionNode(PropertyDefinition property)
+			{
+				_property = property;
+			}
+
+			public override bool InterestingForDynamicDependencyAnalysis => false;
+
+			public override bool HasDynamicDependencies => false;
+
+			public override bool HasConditionalStaticDependencies => false;
+
+			public override bool StaticDependenciesAreComputed => true;
+
+			public override IEnumerable<CombinedDependencyListEntry>? GetConditionalStaticDependencies (NodeFactory context) => null;
+
+			public override IEnumerable<DependencyListEntry>? GetStaticDependencies (NodeFactory context)
+			{
+				using var propertyScope = context.MarkStep.ScopeStack.PushLocalScope (new MessageOrigin (_property));
```

Comment:
I haven't looked deeply into it, but since it's basically global state for the MarkStep execution, and IIRC there are a places that don't take a MessageOrigin as a parameter but reference ScopeStack.CurrentScope.Origin, I think it makes sense to leave it for now. This we'll also have a few DependencyNodes to test the behavior of the replacement. We could add a parameter for each of these places, and make all the existing optional parameters mandatory, but I think that should be in a follow-up or separate PR.

I agree we should eventually get rid of the ScopeStack concept, though. I'm working on a separate PR to translate the MessageOrigin concept to the dependency analysis framework, then we can start removing its usage. It depends on the EdgeVisitor in https://github.com/dotnet/runtime/pull/101949, and my current plan is to do something similar to the ITracingNode from that PR.

Reply:
I don't remember all the details, but conceptually there should be no need for the scope stack if we correctly separate concerns:
- Stuff which needs to report warnings on behalf of the "callsite" should be done "on-call", so we would still keep things like `MarkProperty`, but it would produce a dependency node list which the caller needs to add to the graph. This is similar to the concept of MetadataManager in NativeAOT - basically all the methods on that class act like "Mark" methods in the trimmer - they compute the "on-call" information and delay the rest by returning a dependnecy node list.
- Stuff which is only about the property itself (like attribute processing for example) should happen inside the `PropertyNode` in its `GetDependencies` - this doesn't need scope stack, as the scope is always the property itself. For example `MarkProperty` is all "delayed":
```csharp
		protected internal void MarkProperty (PropertyDefinition prop, in DependencyInfo reason)
		{
			if (!Annotations.MarkProcessed (prop, reason))
				return;

            // Anything after this line can be delayed as it's not per callsite, but per target - we push the property 
            // on the stack, so everything is blamed on the property itself - and thus can be done delayed.
            //  Anything before this should be done per-call - in this case, there's nothing done per-call...
			using var propertyScope = ScopeStack.PushLocalScope (new MessageOrigin (prop));

			// Consider making this more similar to MarkEvent method?
			MarkCustomAttributes (prop, new DependencyInfo (DependencyKind.CustomAttribute, prop));
			DoAdditionalPropertyProcessing (prop);
		}
```

Reply:
This is similar in all the `MarkSomething` methods - they eventually push onto the scope stack - and that's the point which we would split and delay through a node in the graph.

Reply:
Finally - we would pass the scope as a parameter to the methods which would need it and which are called from these. But there should not be that many once we're done with these changes.

Reply:
One downside of solving the `MessageOrigin` problem with up-front marking is that the final graph won't have all the dependencies since there may be transitive dependencies in the up-front marking. It's probably not a huge deal, but if we can figure out a way to report warnings that require the `MessageOrigin` in an `EdgeVisitor`, and are able to avoid up-front marking, we could have the full dependency graph. That could be worked out later if we find it valuable, though.

Regardless, before we can return anything from `GetStaticDependencies`, we need to add the dependency tracing `EdgeVisitor` in https://github.com/dotnet/runtime/pull/101949 to log the dependencies properly.

Reply:
Sorry I didn't read the other PR - why do we need something new to report dependencies correctly? Or rather, why doesn't NativeAOT have the same problem?

Reply:
Actually, thinking more about it, if we do up-front marking for each node and mark the node with a reason during the dependency analysis, that PR isn't a prerequisite.